// facilities/allowed.js
let models = require('sql-models');
exports = {
  async get_allowed_facilities(user_id){
    // console.log("user_id",user_id)
    // user_id=52;
    let roles = await models.db.vw_roles.findAll({
      attributes:['role_code'],
      where: {
        user_id: user_id
      },
      raw: true
    })
    let facilities_options = await models.db.vw_data_types.findAll({
      attributes:['option', 'value'],
      where: {
        user_id: user_id,
        $or: roles,
        data_type: 'facility'
      },
      raw: true
    })
    // console.log("facilities_options",facilities_options)
    let default_option = await models.db.wx_auth_data_types.findAll({
      attributes:[['default_value','option']],
      where: {
        data_type: 'facility'
      },
      raw: true
    })
    let options = facilities_options.concat(default_option)
    // let facilities = []
    let where = {
      $or: []
    }
    for(var i=0;i<options.length;i++){
      let option = options[i].option;
      switch(option){
        case 'none':
          where.$or.push({
            facility_id:{
              lte:0
            }
          })
          break;
        case 'constant':
          where.$or.push({
            facility_id:options[i].value
          })
          break;
        case 'all_self_le':
          let self_le = await models.db.employees.findAll({
            attributes:['le_id'],
            where:{
              user_id: user_id
            },
            raw:true
          })
          where.$or.push({
            $or: self_le
          })
          break;
      }
    }
    // console.log("where",  where)
    return where;
  }
};

module.exports = exports;
