// organisations/org-structure.js
let models = require('sql-models');
let sequelize = require("sequelize");


let orgStr = {

    async headPosition({le_id}) {
        return await models.db.sequelize.query(`SELECT * from org_positions where parent_pos_id is NULL AND le_id = ${le_id}`, {type: sequelize.QueryTypes.SELECT});
    },

    async childPositionOfPosition({pos_id, bu_id = null}){
        return await models.db.org_positions.findAll({
            where: {
                parent_pos_id: pos_id,
                bu_id: bu_id //of position
            },
            raw: true
        });
    },

    async childBuOfPosition({pos_id}){
        return await models.db.org_business_units.findAll({
            where: {
                head_pos_id: pos_id
            },
            raw: true
        });
    },

    async childPositionOfBu({head_pos_id, bu_id}){
        return await models.db.org_positions.findAll({
            where: {
                parent_pos_id: head_pos_id,
                bu_id: bu_id
            }
        });
    },

    async createPosition({le_id, bu_id, designation, description, parent_pos_id, pos_name, is_active = true}){
        pos_name = pos_name ? pos_name : designation;
        return models.db.org_positions.create({
            le_id, bu_id, designation, description, parent_pos_id, pos_name, is_active
        })
    },

    async createBU({bu_name, description, le_id, type_id, head_pos_id}){
        return models.db.org_business_units.create({
            bu_name, description, le_id, type_id, head_pos_id
        })
    }
};

module.exports = orgStr;
