// _includes/cassandra/cs_db.js
var models = require('express-cassandra');
var path = __dirname;
console.log('__dirname ', __dirname);

var connection_cassandra;
exports.cass_connection = function () {
    //104.155.216.207
    if (!connection_cassandra) {
        console.log('============================Starting the connection==============================================', connection_cassandra);
        models.setDirectory(path + '/models').bind(
            {
                clientOptions: {
                    contactPoints: ['127.0.0.1'],
                    protocolOptions: {port: 9042},
                    keyspace: 'workexv0',
                    queryOptions: {consistency: models.consistencies.one}
                    // authProvider: new models.driver.auth.DsePlainTextAuthProvider('cassandra', 'jXvB15df')
                },
                ormOptions: {
                    //If your keyspace doesn't exist it will be created automatically
                    //using the default replication strategy provided here.
                    defaultReplicationStrategy: {
                        class: 'SimpleStrategy',
                        replication_factor: 1
                    },
                    migration: 'safe',
                    createKeyspace: true
                }
            },
            function (err) {
                if (err) console.log(err.message);
                else {
                    // console.log(models.modelInstance.Organisations._properties.schema.fields);
                    // console.log('Models Directory', models.directory);
                    // findPerson();
                    console.log('======================================I am connected======================================', models.timeuuid());
                    // findOrganisation();
                    // organisations();
                }
            }
        );
    }
    connection_cassandra = models;
    return connection_cassandra;
}


function findPerson() {
    models.instance.Person.findOne({name: 'Mary'}, function (err, john) {
        if (err) {
            console.log(err);
            return;
        }
        //Note that returned variable john here is an instance of your model,
        //so you can also do john.delete(), john.save() type operations on the instance.
        console.log('Found ' + john.name + ' to be ' + john.age + ' years old!');
    });
}

function findOrganisation() {
    models.instance.Organisations.findOne({
        org_id: models.timeuuidFromString('fe615100-0b96-11e7-90c0-d3f48f3ff6b5')
    }, function (err, result) {
        if (err) {
            console.log(err);
            return;
        }
        console.log(result.org_name);
        // result = result[0];
        // console.log('Found ' + result.org_id + ' and the head ' + result.org_head + ' is awesome!');
    });
}


// module.exports = {
//     cassandra: models,
//     router: router,
//     object: cass,
//     cass_connection: cass_connection
// }
