// _includes/cassandra/models/OrganisationEmployeesModel.js
/**
 * Created by mysterio86 on 21/3/17.
 */

//     org_id TIMEUUID,
//     ou_ent_id TIMEUUID,
//     ou_ent_name VARCHAR,
//     ou_pos_id TIMEUUID,
//     ou_pos_name VARCHAR,
//     emp_id TIMEUUID,
//     emp_parent_id TIMEUUID,
//     emp_id_alias VARCHAR,
//     user_id TIMEUUID,
//     user_full_name VARCHAR,
//     emp_view_obj MAP<VARCHAR, VARCHAR>,
//     user_img BLOB,
//     is_active BOOLEAN,
//     created_by TIMEUUID,
//     created_on TIMESTAMP,
//     updated_by TIMEUUID,
//     updated_on TIMESTAMP,
//     deleted_by TIMEUUID,
//     deleted_on TIMESTAMP,
//     PRIMARY KEY ((org_id, ou_ent_id), ou_pos_id)

module.exports = {
    fields: {
        org_id: "timeuuid",
        ou_ent_id: "timeuuid",
        ou_ent_name: "text",
        ou_pos_id: "timeuuid",
        ou_pos_name: "text",
        emp_id: {
            type: "timeuuid",
            default: {"$db_function": "now()"}
        },
        emp_boss_id: "timeuuid",
        emp_boss_name: "text",
        emp_id_alias: "text",
        user_id: "timeuuid",
        user_full_name: "text",
        emp_view_obj: {
            type: "map",
            typeDef: "<text, text >"
        },
        user_img: "blob",
        is_active: {
            type: "boolean", default: true
        },
        created_by: {type: "timeuuid"},
        created_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        },
        deleted_by: {type: "timeuuid"},
        deleted_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        },
        updated_by: {type: "timeuuid"},
        updated_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        }
    },
    key: [["org_id", "ou_ent_id"], "ou_pos_id", "user_id"],
    table_name: "organisation_employees"
}
