// _includes/cassandra/models/OrganisationUnitPositionsModel.js
/**
 * Created by mysterio86 on 20/3/17.
 */

// org_id TIMEUUID,
//     org_ou_id TIMEUUID,
//     org_ou_code VARCHAR,
//     org_ou_pos_id TIMEUUID,
//     org_ou_pos_name VARCHAR,
//     org_ou_pos_count INT,
//     org_ou_par_pos_id TIMEUUID,
//     access_level INT,
//     is_active BOOLEAN,
//     org_name VARCHAR,
//     created_by TIMEUUID,
//     created_on TIMESTAMP,
//     updated_by TIMEUUID,
//     updated_on TIMESTAMP,
//     deleted_by TIMEUUID,
//     deleted_on TIMESTAMP,

module.exports = {
    fields: {
        org_id: {
            type: "timeuuid",
            rule: {
                required: true
            }
        },
        ou_id: {
            type: "timeuuid",
            rule: {
                required: true
            }
        },
        ou_code: "text",
        ou_pos_id: {
            type: "timeuuid",
            default: {"$db_function": "now()"}
        },
        ou_pos_name: "text",
        ou_pos_count: "int",
        ou_par_pos_id: {
            type: "timeuuid",
            rule: {
                required: true
            }
        },
        access_level: "int",
        is_active: {
            type: "boolean",
            default: true
        },
        org_name: "text",
        created_by: {type: "timeuuid"},
        created_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        },
        deleted_by: {type: "timeuuid"},
        deleted_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        },
        updated_by: {type: "timeuuid"},
        updated_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        },
    },
    key: [["org_id"], "ou_id", "ou_pos_name"],
    table_name: "organisation_unit_positions"

}
