// _includes/cassandra/models/OrganisationStructureModel.js
/**
 * Created by mysterio86 on 20/3/17.
 */

// DROP TABLE IF EXISTS workexv0.organisation_structure;
// CREATE TABLE IF NOT EXISTS workexv0.organisation_structure(
//     org_id TIMEUUID,
//     org_ou_id TIMEUUID,
//     org_ou_level TINYINT,
//     org_ou_ent_id TIMEUUID,
//     org_ou_ent_name VARCHAR,
//     org_ou_ent_pid TIMEUUID,
//     org_ou_ent_pname VARCHAR,
//     org_ou_ent_desc VARCHAR,
//     org_ou_ent_head_pos VARCHAR,
//     org_ou_ent_head SET<VARCHAR>,
//     org_ou_ent_no_pos INT,
//     org_ou_ent_alias VARCHAR,
//     is_active BOOLEAN,
//     org_name VARCHAR,
//     org_ou_code VARCHAR,
//     created_by TIMEUUID,
//     created_on TIMESTAMP,
//     updated_by TIMEUUID,
//     updated_on TIMESTAMP,
//     deleted_by TIMEUUID,
//     deleted_on TIMESTAMP,
//     PRIMARY KEY ((org_id), org_ou_id, org_ou_ent_id)
// );

module.exports = {
    fields: {
        org_id: {
            type: 'timeuuid',
            rule: {
                required: true
            }
        },
        ou_id: {
            type: 'timeuuid',
            rule: {
                required: true
            }
        },
        ou_level: {
            type: "tinyint",
            rule: {
                required: true
            }
        },
        ou_ent_id: {
            type: 'timeuuid',
            default: {'$db_function': "now()"}
        },
        ou_ent_code: {
            type: 'text',
            rule: {
                required: true
            }
        },
        ou_ent_name: {
            type: 'text',
            rule: {
                required: true
            }
        },
        ou_ent_pid: {
            type: "timeuuid",
            rule: {
                required: true
            }
        },
        ou_ent_pname: "text",
        ou_ent_desc: "text",
        ou_ent_head_pos: "text",
        ou_ent_head: {
            type: "set",
            typeDef: "<varchar>"
        },
        ou_ent_no_pos: "int",
        is_active: {
            type: "boolean",
            default: true
        },
        org_name: "text",
        ou_code: "text",
        created_by: {type: "timeuuid"},
        created_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        },
        deleted_by: {type: "timeuuid"},
        deleted_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        },
        updated_by: {type: "timeuuid"},
        updated_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        }

    },
    key: [["org_id"], "ou_id", "ou_ent_code"],
    table_name: "organisation_structure"
}
