// _includes/cassandra/models/OrganisationUnitTypesModel.js
/**
 * Created by mysterio86 on 20/3/17.
 */

// DROP TABLE IF EXISTS workexv0.organisation_unit_types;
// CREATE TABLE IF NOT EXISTS workexv0.organisation_unit_types(
//     org_id TIMEUUID,
//     org_ou_id TIMEUUID,
//     org_ou_key VARCHAR,
//     org_ou_name VARCHAR,
//     org_ou_description VARCHAR,
//     org_name VARCHAR,
//     is_active BOOLEAN,
//     created_by TIMEUUID,
//     created_on TIMESTAMP,
//     updated_by TIMEUUID,
//     updated_on TIMESTAMP,
//     deleted_by TIMEUUID,
//     deleted_on TIMESTAMP,
//     PRIMARY KEY ((org_id),  org_ou_id)
// );

let validation = require('./ValidationFns');
let options = {min: 3, max: undefined};

module.exports = {
    fields: {
        org_id: {type: "timeuuid"},
        ou_id: {
            type: 'timeuuid',
            default: {'$db_function': 'now()'}
        },
        ou_code: {
            type: "text",
            rule: {
                required: true,
                validators: [
                    {
                        validator: validation.isEmpty,
                        message: function (value) {
                            return 'Please provide the organisation unit name';
                        }
                    },
                    {
                        validator: function (value) {
                            return validation.isLength(value, options);
                        },
                        message: function (value) {
                            return 'Organisation unit code should be atleast ' + options.min + ' alphabets long!'
                        }
                    },
                    {
                        validator: validation.containsSpace,
                        message: function (value) {
                            return 'Organisation unit code should not contain space'
                        }
                    }
                ]
            }
        },
        ou_name: {
            type: "text",
            rule: {
                required: true,
                validators: [
                    {
                        validator: validation.isEmpty,
                        message: function (value) {
                            return 'Please provide the organisation unit name';
                        }
                    },
                    {
                        validator: function (value) {
                            return validation.isLength(value, options);
                        },
                        message: function (value) {
                            return 'Organisation unit code should be atleast ' + options.min + ' alphabets long!'
                        }
                    }
                ]
            }
        },
        ou_description: {type: "text"},
        org_name: {
            type: "text",
            rule: {
                required: true,
                validators: [
                    {
                        validator: validation.isEmpty,
                        message: function (value) {
                            return 'Please provide the organisation unit name';
                        }
                    },
                    {
                        validator: function (value) {
                            return validation.isLength(value, options);
                        },
                        message: function (value) {
                            return 'Organisation unit code should be atleast ' + options.min + ' alphabets long!'
                        }
                    }
                ]
            }
        },
        is_active: {
            type: "boolean",
            rule: {
                required: true
            },
            default: true
        },
        created_by: {type: "timeuuid"},
        created_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        },
        deleted_by: {type: "timeuuid"},
        deleted_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        },
        updated_by: {type: "timeuuid"},
        updated_on: {
            type: "timestamp",
            default: {"$db_function": "toTimestamp(now())"}
        }
    },
    key: [["org_id"], "ou_code"],
    table_name: "organisation_unit_types",
    materialized_views: {
        ous_by_status: {
            select: ["is_active", "org_name", "org_id", "ou_id", "ou_code", "ou_name"],
            key : ["is_active", "org_id", "ou_code"],
        }
    },
}
