// auth-module/otp.js
let speakeasy = require("speakeasy");
let sms_service = require('sms_service');
// import * as _ from "lodash";

let OTP_VALIDATION_TIME = 60 * 5; // in seconds
let OTP_ENCODING = 'base32'; // in seconds

let otp = {
    getNewOTP() {
        // base 32 encoded user secret key
        let secret = speakeasy.generateSecret({length: 20});
        // otp code
        let code = speakeasy.totp({
            secret: secret.base32,
            encoding: OTP_ENCODING,
            step: OTP_VALIDATION_TIME
        });
        return {secret: secret.base32, code};
    },
    validateOtp(code, secret) {
        let status = speakeasy.totp.verify({
            secret: secret,
            encoding: OTP_ENCODING,
            token: code,
            step: OTP_VALIDATION_TIME
        });
        return status;
    },
    sendLoginOTPonSMS(mobile, otp){
        let msg = 'Use ' + otp + ' as your login OTP. OTP is confidential. Sharing it with anyone gives them full access to your Workex Account.';
        sms_service.send_sms(mobile, msg);
    }
}
exports.otp = otp;
