// auth-module/UserFunction.js
let models = require('sql-models');

let FunctionList = {

    async getAllUser(attributes, paging){
        let {currentPage, pageSize} = paging;
        // console.log("paging => ", paging);
        return await models.db.auth_users.findAll({
            limit : pageSize ? parseInt(`${pageSize}`) : undefined,
            offset: currentPage ? parseInt(currentPage) * parseInt(pageSize) : undefined,
        });
    },

    async getPageCount(pageSize){
        let count = await models.db.auth_users.count();
        return _.ceil(count / pageSize);
    },

    async findModules(object){
     return await models.db.sequelize.query(object);
   },

   async getEmployee(object){

     return await models.db.employees.find({
     where: {user_id:object}

    })
    },

    async createNewUser(object){
        return await models.db.auth_users.create(object);
    },

    async createNewUsers(object){
        return await models.db.auth_users.bulkCreate(object);
    },

    async getUserFromKey(key) {
        return await models.db.auth_users.findOne({
            where: {
                $or: [
                    {username: key},
                    {mobile: key},
                    {work_email: key},
                ]
            }
        });
    }

};

exports.fun = FunctionList;
