// sql-models/client_vendor.js
/* jshint indent: 2 */

module.exports = function(sequelize, DataTypes) {
  return sequelize.define('client_vendor', {
    id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      primaryKey: true,
      autoIncrement: true
    },
    relation_type: {
      type: DataTypes.ENUM('pe-sa'),
      allowNull: true
    },
    le_id_a: {
      type: DataTypes.INTEGER(11),
      allowNull: true,
      references: {
        model: 'org_legal_entities',
        key: 'le_id'
      }
    },
    le_id_b: {
      type: DataTypes.INTEGER(11),
      allowNull: true,
      references: {
        model: 'org_legal_entities',
        key: 'le_id'
      }
    },
    is_active: {
      type: DataTypes.INTEGER(1),
      allowNull: true,
      defaultValue: '1'
    },
    created_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    created_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    updated_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    updated_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    deleted_by: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    deleted_on: {
      type: DataTypes.TIME,
      allowNull: true
    }
  }, {
    tableName: 'client_vendor'
  });
};
