// sql-models/emp_geolocation.js
/* jshint indent: 2 */

module.exports = function(sequelize, DataTypes) {
  return sequelize.define('emp_geolocation', {
    document_id: {
      type: DataTypes.BIGINT,
      allowNull: false,
      primaryKey: true,
      references: {
        model: 'employee_attendance',
        key: 'id'
      }
    },
    document_type: {
      type: DataTypes.ENUM('attendance_checkin','attendance_checkout'),
      allowNull: true,
      defaultValue: 'attendance_checkin'
    },
    description: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    geolat: {
      type: DataTypes.FLOAT,
      allowNull: true
    },
    geolng: {
      type: DataTypes.FLOAT,
      allowNull: true
    },
    is_active: {
      type: DataTypes.INTEGER(1),
      allowNull: false,
      defaultValue: '1'
    },
    created_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    created_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    updated_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    updated_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    deleted_by: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    deleted_on: {
      type: DataTypes.TIME,
      allowNull: true
    }
  }, {
    tableName: 'emp_geolocation'
  });
};
