// sql-models/employees_pe_details.js
/* jshint indent: 2 */

module.exports = function(sequelize, DataTypes) {
  return sequelize.define('employees_pe_details', {
    offroll_id: {
      type: DataTypes.BIGINT,
      allowNull: false,
      primaryKey: true,
      autoIncrement: true
    },
    emp_id: {
      type: DataTypes.BIGINT,
      allowNull: false,
      references: {
        model: 'employees',
        key: 'emp_id'
      }
    },
    le_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      references: {
        model: 'org_legal_entities',
        key: 'le_id'
      }
    },
    dm: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    facility_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    emp_code: {
      type: DataTypes.STRING(200),
      allowNull: true
    },
    pos_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    bu_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    emp_type: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    DOJ: {
      type: DataTypes.DATE,
      allowNull: true
    },
    DOL: {
      type: DataTypes.DATE,
      allowNull: true
    },
    is_active: {
      type: DataTypes.INTEGER(1),
      allowNull: true,
      defaultValue: '1'
    },
    created_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    created_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    updated_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    updated_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    deleted_by: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    deleted_on: {
      type: DataTypes.TIME,
      allowNull: true
    }
  }, {
    tableName: 'employees_pe_details'
  });
};
