// sql-models/employee_attendance.js
/* jshint indent: 2 */

module.exports = function(sequelize, DataTypes) {
  return sequelize.define('employee_attendance', {
    id: {
      type: DataTypes.BIGINT,
      allowNull: false,
      primaryKey: true,
      autoIncrement: true
    },
    employee_id: {
      type: DataTypes.BIGINT,
      allowNull: false,
      references: {
        model: 'employees',
        key: 'emp_id'
      }
    },
    photo_url: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    attendance_date: {
      type: DataTypes.DATE,
      allowNull: true
    },
    attendance_code: {
      type: DataTypes.STRING(3),
      allowNull: true
    },
    description: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    reason_code: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    remark: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    created_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    created_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    updated_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    updated_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    time_in: {
      type: DataTypes.TIME,
      allowNull: true
    },
    time_out: {
      type: DataTypes.TIME,
      allowNull: true
    },
    checkout_photo_url: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    checkout_remark: {
      type: DataTypes.STRING(100),
      allowNull: true
    }
  }, {
    tableName: 'employee_attendance'
  });
};
