// sql-models/exit_requests.js
/* jshint indent: 2 */

module.exports = function(sequelize, DataTypes) {
  return sequelize.define('exit_requests', {
    id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      primaryKey: true,
      autoIncrement: true
    },
    emp_code: {
      type: DataTypes.STRING(45),
      allowNull: false
    },
    dol: {
      type: DataTypes.DATE,
      allowNull: false
    },
    leaving_reason: {
      type: DataTypes.STRING(250),
      allowNull: true
    },
    resignation_reason: {
      type: DataTypes.STRING(250),
      allowNull: true
    },
    notice_period_given: {
      type: DataTypes.INTEGER(4),
      allowNull: true
    },
    notice_period_waived: {
      type: DataTypes.INTEGER(4),
      allowNull: true
    },
    created_by: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    created_on: {
      type: DataTypes.DATE,
      allowNull: true
    },
    updated_by: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    updated_on: {
      type: DataTypes.DATE,
      allowNull: true
    },
    is_revoked: {
      type: DataTypes.INTEGER(4),
      allowNull: true
    },
    current_status: {
      type: DataTypes.STRING(45),
      allowNull: true
    }
  }, {
    tableName: 'exit_requests'
  });
};
