// sql-models/olrs.js
/* jshint indent: 2 */

module.exports = function(sequelize, DataTypes) {
  return sequelize.define('olrs', {
    olr_id: {
      type: DataTypes.BIGINT,
      allowNull: false,
      primaryKey: true,
      autoIncrement: true
    },
    pe_le_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      references: {
        model: 'org_legal_entities',
        key: 'le_id'
      }
    },
    pe_dm: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    pe_facility_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    pe_pos_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    pe_bu_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    designation: {
      type: DataTypes.STRING(45),
      allowNull: true
    },
    is_offroll: {
      type: DataTypes.INTEGER(1),
      allowNull: true
    },
    emp_type: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    DOJ: {
      type: DataTypes.DATE,
      allowNull: true
    },
    permanent_address: {
      type: DataTypes.BIGINT,
      allowNull: true
    },
    correspondence_address: {
      type: DataTypes.BIGINT,
      allowNull: true
    },
    DOB: {
      type: DataTypes.DATE,
      allowNull: true
    },
    fname: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    lname: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    mobile: {
      type: DataTypes.STRING(15),
      allowNull: true
    },
    per_email: {
      type: DataTypes.STRING(150),
      allowNull: true
    },
    aadhar_no: {
      type: DataTypes.STRING(16),
      allowNull: true
    },
    aadhar_file_url: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    base_location_city: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    base_location_state: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    base_location_pincode: {
      type: DataTypes.STRING(20),
      allowNull: true
    },
    gender: {
      type: DataTypes.ENUM('male','female'),
      allowNull: true
    },
    fathers_name: {
      type: DataTypes.STRING(200),
      allowNull: true
    },
    mothers_name: {
      type: DataTypes.STRING(200),
      allowNull: true
    },
    marital_status: {
      type: DataTypes.ENUM('single','married','others'),
      allowNull: true
    },
    spouse_name: {
      type: DataTypes.STRING(200),
      allowNull: true
    },
    payroll_le_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false
    },
    payroll_dm: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    payroll_facility_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    payroll_pos_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    payroll_bu_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    payroll_emp_type: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    created_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    created_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    updated_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    updated_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    deleted_by: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    deleted_on: {
      type: DataTypes.TIME,
      allowNull: true
    },
    current_status: {
      type: DataTypes.STRING(45),
      allowNull: false,
      defaultValue: 'Draft'
    }
  }, {
    tableName: 'olrs'
  });
};
