// sql-models/org_legal_entities.js
/* jshint indent: 2 */

module.exports = function(sequelize, DataTypes) {
  return sequelize.define('org_legal_entities', {
    le_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      primaryKey: true,
      autoIncrement: true
    },
    org_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      references: {
        model: 'organisations',
        key: 'org_id'
      }
    },
    le_name: {
      type: DataTypes.STRING(250),
      allowNull: true
    },
    le_code: {
      type: DataTypes.STRING(45),
      allowNull: true
    },
    logo_file: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    description: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    address_line1: {
      type: DataTypes.STRING(500),
      allowNull: false
    },
    address_line2: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    locality: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    landmark: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    city: {
      type: DataTypes.STRING(150),
      allowNull: false
    },
    state: {
      type: DataTypes.STRING(150),
      allowNull: false
    },
    country: {
      type: DataTypes.STRING(150),
      allowNull: false
    },
    pincode: {
      type: DataTypes.STRING(20),
      allowNull: false
    },
    is_active: {
      type: DataTypes.INTEGER(1),
      allowNull: true,
      defaultValue: '1'
    },
    created_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    created_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    updated_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    updated_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    deleted_by: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    deleted_on: {
      type: DataTypes.TIME,
      allowNull: true
    }
  }, {
    tableName: 'org_legal_entities'
  });
};
