// sql-models/org_positions.js
/* jshint indent: 2 */

module.exports = function(sequelize, DataTypes) {
  return sequelize.define('org_positions', {
    pos_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      primaryKey: true,
      autoIncrement: true
    },
    le_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      references: {
        model: 'org_legal_entities',
        key: 'le_id'
      }
    },
    bu_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true,
      references: {
        model: 'org_business_units',
        key: 'bu_id'
      }
    },
    designation: {
      type: DataTypes.STRING(150),
      allowNull: true
    },
    description: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    parent_pos_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true,
      references: {
        model: 'org_positions',
        key: 'pos_id'
      }
    },
    pos_name: {
      type: DataTypes.STRING(150),
      allowNull: true
    },
    is_active: {
      type: DataTypes.INTEGER(1),
      allowNull: true,
      defaultValue: '1'
    },
    created_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    created_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    updated_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    updated_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    deleted_by: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    deleted_on: {
      type: DataTypes.TIME,
      allowNull: true
    }
  }, {
    tableName: 'org_positions'
  });
};
