// sql-models/org_business_units.js
/* jshint indent: 2 */

module.exports = function(sequelize, DataTypes) {
  return sequelize.define('org_business_units', {
    bu_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      primaryKey: true,
      autoIncrement: true
    },
    bu_name: {
      type: DataTypes.STRING(250),
      allowNull: true
    },
    description: {
      type: DataTypes.TEXT,
      allowNull: true
    },
    le_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      references: {
        model: 'org_legal_entities',
        key: 'le_id'
      }
    },
    type_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      references: {
        model: 'org_entity_types',
        key: 'type_id'
      }
    },
    head_pos_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      references: {
        model: 'org_positions',
        key: 'pos_id'
      }
    },
    is_active: {
      type: DataTypes.INTEGER(1),
      allowNull: true,
      defaultValue: '1'
    },
    created_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    created_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    updated_by: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '-1'
    },
    updated_on: {
      type: DataTypes.TIME,
      allowNull: false,
      defaultValue: sequelize.literal('CURRENT_TIMESTAMP')
    },
    deleted_by: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    deleted_on: {
      type: DataTypes.TIME,
      allowNull: true
    }
  }, {
    tableName: 'org_business_units'
  });
};
