// sql-models/vw_employees.js
/* jshint indent: 2 */

module.exports = function(sequelize, DataTypes) {
  return sequelize.define('vw_employees', {
    user_id: {
      type: DataTypes.BIGINT,
      allowNull: true,
      defaultValue: '0'
    },
    emp_id: {
      type: DataTypes.BIGINT,
      allowNull: false,
      defaultValue: '0'
    },
    emp_code: {
      type: DataTypes.STRING(45),
      allowNull: false
    },
    pe_emp_code: {
      type: DataTypes.STRING(200),
      allowNull: true
    },
    facility_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true,
      defaultValue: '0'
    },
    facility_name: {
      type: DataTypes.STRING(250),
      allowNull: true
    },
    region: {
      type: DataTypes.STRING(45),
      allowNull: true
    },
    le_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false
    },
    org_name: {
      type: DataTypes.STRING(250),
      allowNull: true
    },
    pe_le_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    fname: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    lname: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    pos_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    pe_pos_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    pe_bu_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    bu_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    is_active: {
      type: DataTypes.INTEGER(1),
      allowNull: true,
      defaultValue: '1'
    },
    full_name: {
      type: DataTypes.STRING(201),
      allowNull: true
    },
    initials: {
      type: DataTypes.STRING(2),
      allowNull: true
    },
    mobile: {
      type: DataTypes.STRING(15),
      allowNull: true
    },
    work_email: {
      type: DataTypes.STRING(150),
      allowNull: true
    },
    designation: {
      type: DataTypes.STRING(45),
      allowNull: true
    },
    base_location_city: {
      type: DataTypes.STRING(100),
      allowNull: true
    },
    DOB: {
      type: DataTypes.STRING(38),
      allowNull: true
    },
    DOJ: {
      type: DataTypes.STRING(38),
      allowNull: true
    },
    DOL: {
      type: DataTypes.STRING(38),
      allowNull: true
    },
    display_pic: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    gender: {
      type: DataTypes.STRING(1),
      allowNull: false,
      defaultValue: ''
    },
    is_offroll: {
      type: DataTypes.STRING(7),
      allowNull: false,
      defaultValue: ''
    },
    op_type: {
      type: DataTypes.STRING(45),
      allowNull: true
    },
    op_subtype: {
      type: DataTypes.STRING(45),
      allowNull: true
    },
    dm_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    },
    pe_dm_id: {
      type: DataTypes.INTEGER(11),
      allowNull: true
    }
  }, {
    tableName: 'vw_employees'
  });
};
