// sql-models/vw_facilities.js
/* jshint indent: 2 */

module.exports = function(sequelize, DataTypes) {
  return sequelize.define('vw_facilities', {
    facility_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false,
      defaultValue: '0'
    },
    le_id: {
      type: DataTypes.INTEGER(11),
      allowNull: false
    },
    logo_file: {
      type: DataTypes.STRING(500),
      allowNull: true
    },
    facility_type: {
      type: DataTypes.STRING(200),
      allowNull: true
    },
    facility_name: {
      type: DataTypes.STRING(250),
      allowNull: false
    },
    address_line1: {
      type: DataTypes.STRING(500),
      allowNull: false
    },
    city: {
      type: DataTypes.STRING(150),
      allowNull: false
    },
    state: {
      type: DataTypes.STRING(150),
      allowNull: false
    },
    pincode: {
      type: DataTypes.STRING(20),
      allowNull: false
    },
    region: {
      type: DataTypes.STRING(45),
      allowNull: true
    },
    location: {
      type: DataTypes.STRING(301),
      allowNull: true
    },
    mgr_contact: {
      type: DataTypes.STRING(15),
      allowNull: true
    },
    mgr_name: {
      type: DataTypes.STRING(201),
      allowNull: true
    }
  }, {
    tableName: 'vw_facilities'
  });
};
