// cassandra-models/index.js
let models         = require('express-cassandra');
let chalk          = require('chalk');
let path           = __dirname;
let config         = require('config');
let casandraConfig = config.get("database").casandra;

/** Expose all models */
// exports = module.exports;

exports.init = function init() {
    return new Promise((resolve, reject) => {
        let {contactPoints, port, keyspace, userName, password} = casandraConfig;

        console.log(chalk.yellow(`connecting to cassandra at host ${contactPoints} on ${port} via keyspace ${keyspace}`));
        models.setDirectory(path).bind({
                clientOptions: {
                    contactPoints,
                    protocolOptions: {port: port},
                    keyspace       : keyspace,
                    queryOptions   : {consistency: models.consistencies.one},
                    authProvider   : new models.driver.auth.DsePlainTextAuthProvider(userName, password)
                },
                ormOptions   : {
                    //If your keyspace doesn't exist it will be created automatically
                    //using the default replication strategy provided here.
                    defaultReplicationStrategy: {
                        class             : 'SimpleStrategy',
                        replication_factor: 1
                    },
                    migration                 : 'safe',
                    createKeyspace            : true,
                    udts                      : {
                        address: {
                            address_line1: "text",
                            address_line2: "text",
                            locality     : "text",
                            landmark     : "text",
                            city         : "text",
                            state        : "text",
                            country      : "text",
                            pincode      : "text"
                        }
                    }
                }
            },
            function (err) {
                if (err) {
                    console.log(chalk.red('***************Error in casandra initialisation ***************', err.message));
                    reject(err);
                }
                else {
                    console.log(chalk.green(`  ✅  connected to cassandra via keyspace ${keyspace}`, models.timeuuid()));
                    console.log(chalk.green(`  ✅  Cassandra Model Initialisation Done`));
                    resolve(true);
                }
            }
        );
    });
};
