// employees/analytics.js
let models    = require('sql-models');
let sequelize = require("sequelize");
let allowed = require("./allowed");

let FunctionList = {

    async getEmployeeCount(auth_user_id){
        // let fac_id = attributes.facility_id || 0;
        let where = await allowed.get_allowed_employees(auth_user_id);
        return await models.db.vw_employees.count({
            where
        });
    },

    async getAgenciesCount(auth_user_id){
      let where = await allowed.get_allowed_employees(auth_user_id);
      return await models.db.vw_employees.aggregate('le_id', 'count',
      {
        where:  where,
        distinct: true,
      })
        // return models.db.vw_facilities.count();
    },

    async getNewJoineesCount(auth_user_id){
      var TodayDate = new Date();
      console.log(TodayDate.getMonth())
      let where = await allowed.get_allowed_employees(auth_user_id);
      return await models.db.vw_employees.count({
          where:{
            $and: [
              where,
              sequelize.where(sequelize.fn('MONTH', sequelize.col('DOJ')), TodayDate.getMonth() + 1 )
            ]
          }
      });
    },

    async getExitsCount(auth_user_id){
      var TodayDate = new Date();
      console.log(TodayDate.getMonth())
      let where = await allowed.get_allowed_employees(auth_user_id);
      return await models.db.vw_employees.count({
          where:{
            $and: [
              where,
              sequelize.where(sequelize.fn('MONTH', sequelize.col('DOL')), TodayDate.getMonth() + 1 )
            ]
          }
      });
    }
};

exports.fun = FunctionList;
