// ui-schema/constants.js
/** WARNING **/
/**
 * Before Making any changes this file, make sure you change those values into Server App also.
 * This is common file shared between both the projects
 */
module.exports = {

    WebHookConstants   : {
        METHOD     : 'GET',
        GET_METHOD : 'GET',
        POST_METHOD: 'POST'
    },
    WxComponents       : {
        LIST         : 'list',
        LIST_ITEM    : 'ListItem',
        MULTI_SELECT : 'multi_select',
        MULTI_SELECT_DROP_DOWN : 'multi_select_drop_down',
        ARRAY        : 'array',
        LINK         : 'link',
        AUTO_COMPLETE: 'auto_complete',
        RADIO        : 'radio',
        DROP_DOWN    : 'drop_down',
        POPOVER      : 'popover',

        BUTTON_FLAT           : 'flat_button',
        BUTTON_RAISED         : 'raised_button',
        BUTTON_ICON           : 'icon_button',
        BUTTON_FAB            : 'floating_action_button',
        BUTTON_UPLOAD         : 'button_upload',
        TOGGLE                : 'toggle',
        IMAGE_PREVIEW_UPLOADER: 'image_preview_uploader',

        TEXT     : 'text',
        TEXT_AREA: 'text_area',


        CODE         : 'code',

        DATE: 'date',
        TIME: 'time',

        ACTIONS : 'actions',
        ACTION_ITEM : 'action_item',

        AVATAR : 'avatar',
        ICON   : 'icon',
        BADGE  : 'badge',
        DIVIDER: 'divider',

        CHECKBOX: 'checkbox',
        SLIDER  : 'slider',
        DIALOG  : 'dialog',

        HEADER: 'header',
        SPAN  : 'span',


        TAB : 'tab',
        TABS: 'tabs',

        SECTION          : 'section',
        CARD_STACK       : 'card_stack',
        PAGE             : 'page',
        MENU             : 'menu',
        MENU_ITEM        : 'menu_item',
        TABLE            : 'table',
        PAPER            : 'paper',
        VERTICAL_STACK   : 'vertical_stack',
        ROW_COL_CONTAINER: 'row_col_container',
        FORM             : 'form',
        LIST_CONTAINER   : 'list_container',
        DEFAULT_TEMPLATE : 'default_template',
        OVERLAY_TEMPLATE : 'overlay_template',
        FETCH_CONTAINER   : 'fetch_container',
    },
    WxElementsConstants: {
        WxTextFieldConstant      : {
            DISABLED: false,
            TYPE    : 'text',
            FULL    : false
        },
        WxTextAreaConstant       : {
            ROWS    : 2,
            DISABLED: false,
            TYPE    : 'text',
            FULL    : false
        },
        WxCheckBoxConstant       : {
            LABEL_POSITION: 'right',
            DISABLED      : false,
            CHECKED       : false
        },
        WxDatePickerConstant     : {
            DISABLED: false
        },
        WxToggleConstant         : {
            DISABLED      : false,
            VALUE         : false,
            LABEL_POSITION: 'right'
        },
        WxSliderConstant         : {
            DISABLED: false,
            VALUE   : 0
        },
        WxFlatButtonConstant     : {
            DISABLED      : false,
            LABEL_POSITION: 'left'
        },
        WxRaisedButtonConstant   : {
            DISABLED      : false,
            LABEL_POSITION: 'left'
        },
        WxFABButtonConstant      : {
            DISABLED: false,
            MINI    : false
        },
        WxIconButtonConstant     : {
            DISABLED: false
        },
        WxRowColContainerConstant: {
            COLUMN_COUNT: 4
        },
        WxTableConstant          : {
            SHOW_CHECK_BOX: false
        },
    }

};
