// ui-schema/index.js
/**
 * Created by Kartik Agarwal.
 */
module.exports = {

    constants  : require('./constants'),
    time_picker: require('./lib/TimePicker'),
    date_picker: require('./lib/DatePicker'),

    actions     : require('./lib/Actions'),
    actions_item: require('./lib/ActionsItem'),

    auto_complete: require('./lib/AutoComplete'),
    drop_down    : require('./lib/DropDown'),
    multi_select : require('./lib/MultiSelect'),
    multi_select_drop_down : require('./lib/MultiSelectDropDown'),
    array        : require('./lib/Array'),
    icon         : require('./lib/Icon'),

    radio_button_group: require('./lib/RadioButtonGroup'),
    check_box         : require('./lib/CheckBox'),
    toggle            : require('./lib/Toggle'),


    tabs: require('./lib/Tabs'),
    tab : require('./lib/Tab'),

    button_flat  : require('./lib/ButtonFlat'),
    button_raised: require('./lib/ButtonRaised'),
    button_fab   : require('./lib/ButtonFab'),
    button_icon  : require('./lib/ButtonIcon'),
    button_upload: require('./lib/ButtonUpload'),

    avatar: require('./lib/Avatar'),
    badge : require('./lib/Badge'),
    span  : require('./lib/Span'),
    link  : require('./lib/Link'),

    code: require('./lib/Code'),

    slider: require('./lib/Slider'),
    paper : require('./lib/Paper'),

    text_area : require('./lib/TextArea'),
    text_field: require('./lib/TextField'),

    section: require('./lib/Section'),
    divider: require('./lib/Divider'),

    header: require('./lib/Header'),
    page  : require('./lib/Page'),
    form  : require('./lib/Form'),

    list     : require('./lib/List'),
    list_item: require('./lib/ListItem'),

    popover: require('./lib/Popover'),
    menu   : require('./lib/Menu'),

    table     : require('./lib/Table'),
    card_stack: require('./lib/CardStack'),

    vertical_stack        : require('./lib/VerticalStack'),
    default_template        : require('./lib/DefaultTemplate'),
    overlay_template        : require('./lib/OverlayTemplate'),
    row_col_container     : require('./lib/RowColContainer'),
    image_preview_uploader: require('./lib/ImagePreviewUploader'),

    fetch_container: require('./lib/FetchContainer'),
    form_components: require('./lib/FormComponent.js')
};
