// cassandra-modules/auth/User.js
// let models = require('cassandra-models').models();
let models = require('express-cassandra');

class User {

    constructor(param) {
        this.user = param || {};
        this.user_cred = param || {};
    }


    getUserFromKey(key) {
        return new Promise((resolve, reject) => {
            models.instance.AuthUserCredentials.findOne({key}, (err, result) => {
                if (err)
                    reject(err);
                else {
                    this.user_cred = Object.assign({}, this.user_cred, result ? result.toJSON() : {});
                    resolve(this.user_cred);
                }
            });
        })
    }

    getUserByName(userName) {
        return new Promise((resolve, reject) => {
            models.instance.AuthUsers.findOne({user_name: userName}, (err, result) => {
                if (err)
                    reject(err);
                else {
                    this.user = Object.assign({}, this.user, result.toJSON());
                    resolve(this.user);
                }
            });
        });
    }

    getUserByID(user_id) {
        // user_id = 'd507bcc7-258d-11e7-971e-036b8157e331'
        if (typeof  user_id === 'string') {
            user_id = models.timeuuidFromString(user_id);
        }

        return new Promise((resolve, reject) => {
            models.instance.AuthUsers.findOne({user_id: user_id}, (err, result) => {
                if (err)
                    reject(err);
                else {
                    this.user = Object.assign({}, this.user, result.toJSON());
                    console.log(this.user)
                    resolve(this.user);
                }
            });
        });
    }

    // createUser(obj) {
    //     return new Promise((resolve, reject) => {
    //         let data = {
    //             firstname : obj.firstname,
    //             lastname  : obj.lastname,
    //             mobile    : obj.mobile,
    //             user_id   : models.timeuuid(),
    //             created_by: obj.created_by,
    //             updated_by: obj.created_by
    //         };
    //         let user = new models.instance.Users(data);
    //         user.save((err) => {
    //             if (err)
    //                 reject(err);
    //             else {
    //                 this.user = user.toJSON();
    //                 resolve(this.user);
    //             }
    //         });
    //     });
    // }
    //
    // createUserCredentials(obj) {
    //     return new Promise((resolve, reject) => {
    //         let data      = {
    //             username  : obj.username,
    //             password  : obj.password,
    //             user_id   : this.user.user_id,
    //             created_by: this.user.created_by,
    //             updated_by: this.user.updated_by
    //         };
    //         let user_cred = new models.instance.UserCredential(data);
    //         user_cred.save((err) => {
    //             if (err)
    //                 reject(err);
    //             else {
    //                 this.user.username = user_cred.username;
    //                 this.user.password = user_cred.password;
    //                 resolve(this.user);
    //             }
    //         });
    //     });
    // }
    //
    // createMobileCredentials(obj) {
    //     return new Promise((resolve, reject) => {
    //         let data = {
    //             username  : obj.username,
    //             password  : obj.password,
    //             user_id   : this.user.user_id,
    //             created_by: this.user.created_by,
    //             updated_by: this.user.updated_by
    //         };
    //
    //         let user_cred = new models.instance.UserCredential(data);
    //         user_cred.save((err) => {
    //             if (err)
    //                 reject(err);
    //             else {
    //                 this.user.username = user_cred.username;
    //                 this.user.password = user_cred.password;
    //                 resolve(this.user);
    //             }
    //         });
    //     });
    // }
}

exports.userClass = User;
