// cassandra-models/employees/EmployeesModel.js
let TimeStampBase = require('../base/TimeStampBase');
let Schema        = require('../utils/Schema');


let Employees = new Schema(TimeStampBase, {
    fields    : {
        le_id                 : "timeuuid",
        user_id               : "timeuuid",
        aadhar_file_url       : "text",
        aadhar_no             : "text",
        base_location_city    : "text",
        base_location_pincode : "text",
        base_location_state   : "text",
        correspondence_address: {
            type   : "frozen",
            typeDef: "<address>"
        },
        permanent_address     : {
            type   : "frozen",
            typeDef: "<address>"
        },
        display_name          : "text",
        dob                   : "date",
        doj                   : "date",
        dol                   : "date",
        emp_id                : "timeuuid",
        emp_type              : "text",
        facility_id           : "timeuuid",
        fathers_name          : "text",
        fname                 : "text",
        gender                : "text",
        is_active             : "boolean",
        is_offroll            : "boolean",
        lname                 : "text",
        marital_status        : "text",
        mobile                : {
            type   : "set",
            typeDef: "<bigint>"
        },
        per_email             : {
            type   : "set",
            typeDef: "<bigint>"
        },
        mothers_name          : "text",
        org_id                : "timeuuid",
        spouse_name           : "text",
        work_email            : "text",
    },
    key       : [["le_id"], "user_id"],
    table_name: "employees"
});

module.exports = Employees.model();
