// ui-schema/lib/Avatar.js
let {WxComponents, WebHookConstants} =  require('../constants');
let BaseComponent                    = require('./base/BaseComponent');

class Avatar extends BaseComponent {

    static init(param) {
        return new Avatar(param);
    }

    constructor(param) {
        super(param);
        this.config = {'type': WxComponents.AVATAR};
    }

    setBackgroundColor(background) {
        this.config.background = background;
        return this;
    }

    setText(text) {
        this.config.text = text;
        this.config.icon = {
            text
        };
        return this;
    }

    setFontColor(fontColor) {
        this.config.fontColor = fontColor;
        return this;
    }

    setTextColor(fontColor) {
        this.config.fontColor = fontColor;
        return this;
    }

    setBorderRadius(borderRadius) {
        this.config.borderRadius = borderRadius;
        return this;
    }

    setUrl(url) {
        this.config.url  = url;
        this.config.icon = {
            url
        };
        return this;
    }

    setSize(size) {
        this.config.size = size;
        return this;
    }

    setIcon(icon) {
        this.config.oldIcon = icon;
        this.config.icon    = {
            mIcon: icon
        };
        return this;
    }
}

module.exports = Avatar;
