// ui-schema/lib/Badge.js
let {WxComponents}     =  require('../constants');
let BaseComponent      = require('./base/BaseComponent');

class Badge extends BaseComponent {

    static init(param) {
        return new Badge(param);
    }

    constructor(param) {
        super(param);
        this.config = {'type': WxComponents.BADGE};
    }

    setBadgeContent(badgeContent) {
        this.config.badgeContent = badgeContent;
        return this;
    }

    setTextColor(textColor) {
        this.config.textColor = textColor;
        return this;
    }

    setBackgroundColor(backgroundColor) {
        this.config.backgroundColor = backgroundColor;
        return this;
    }

    /**
     * add string or Button icon
     *
     * @param badgeBody
     * @returns {Badge}
     */
    setBadgeBody(badgeBody) {
        if (typeof badgeBody === 'string')
            this.config.badgeBody = badgeBody;
        else
            this.config.badgeBody = badgeBody.render();
        return this;
    }
}

module.exports = Badge;
