// ui-schema/lib/DefaultTemplate.js
let {WxComponents, WebHookConstants} =  require('../constants');
let BaseContainer                    = require('./base/BaseContainer');
let map                              = require('lodash/map');

class DefaultTemplate extends BaseContainer {

    static init(param) {
        return new DefaultTemplate(param);
    }

    constructor(param) {
        super(param);
        this.config = {'type': WxComponents.DEFAULT_TEMPLATE};
    }

    setContentUrl(url, method = WebHookConstants.GET_METHOD, header, body) {
        this.config.content_url = {
            url, method, header, body
        };
        return this;
    }

    setHeader(headerContent) {
        this.config.header = headerContent.render();
        return this;
    }

    addRightAction(contentFactory) {
        if (contentFactory instanceof Array) {
            map(contentFactory, (contentFactoryItem, index) => {
                this.addRightAction(contentFactoryItem)
            })
        } else {
            if (!this.config.action_items)
                this.config.action_items = [];
            try {
                this.config.action_items.push(contentFactory.render());
            } catch (err) {
                console.log(chalk.yellow('Don\'t use render() on addContent'));
                this.config.action_items.push(contentFactory);
            }
        }
        return this;
    };
}

module.exports = DefaultTemplate;
