// ui-schema/lib/Popover.js
let {WxComponents, WebHookConstants} =  require('../constants');
let BaseContainer                    = require('./base/BaseContainer');
let map                              = require('lodash/map');

class Popover extends BaseContainer {

    static init(param) {
        return new Popover(param);
    }

    constructor(param) {
        super(param);
        this.config = {'type': WxComponents.POPOVER};
    }

    setActionElement(actionElement) {
        try {
            this.config.actionElement = actionElement.render();
        }
        catch (err) {
            console.error("WARING : Don't use render method while adding action element in popver ");
            this.config.actionElement = actionElement;
        }
        return this;
    }

    setAnchorOrigin(anchorOrigin = {horizontal: 'left', vertical: 'bottom'}) {
        this.config.anchorOrigin = anchorOrigin;
        return this;
    }

    setAnchorOriginHorizontal(horizontal = 'left') {
        if (!this.config.anchorOrigin)
            this.config.anchorOrigin = {}
        this.config.anchorOrigin.horizontal = horizontal;
        return this;
    }

    setAnchorOriginVertical(vertical = 'bottom') {
        if (!this.config.anchorOrigin)
            this.config.anchorOrigin = {}
        this.config.anchorOrigin.vertical = vertical;
        return this;
    }

    setTargetOrigin(targetOrigin = {horizontal: 'left', vertical: 'top'}) {
        this.config.targetOrigin = targetOrigin;
        return this;
    }

    setTargetOriginHorizontal(horizontal = 'left') {
        if (!this.config.anchorOrigin)
            this.config.anchorOrigin = {}
        this.config.targetOrigin.horizontal = horizontal;
        return this;
    }

    setTargetOriginVertical(vertical = 'top') {
        if (!this.config.anchorOrigin)
            this.config.anchorOrigin = {}
        this.config.targetOrigin.vertical = vertical;
        return this;
    }

}

module.exports = Popover;
