// ui-schema/lib/RowColContainer.js
let {WxComponents}              =  require('../constants');
let {WxRowColContainerConstant} = require('../constants').WxElementsConstants;
let BaseContainer               = require('./base/BaseContainer');

class RowColContainer extends BaseContainer {

    static init(param) {
        return new RowColContainer(param);
    }

    constructor(param) {
        super(param);
        this.config = {
            'type'     : WxComponents.ROW_COL_CONTAINER,
            columnCount: WxRowColContainerConstant.COLUMN_COUNT
        };
    }

    setColumnCount(columnCount) {
        this.config.columnCount = columnCount;
        return this;
    }

    setRowSpacing(rowSpacing) {
        this.config.rowSpacing = rowSpacing;
        return this;
    }

    setRowStart(start) {
        this.config.start = start;
        return this;
    }


    setRowReverse(reverse) {
        this.config.reverse = reverse;
        return this;
    }

    setRowCenter(center) {
        this.config.center = center;
        return this;
    }

    setRowEnd(end) {
        this.config.end = end;
        return this;
    }

    setRowTop(top) {
        this.config.top = top;
        return this;
    }

    setRowMiddle(middle) {
        this.config.middle = middle;
        return this;
    }

    setRowBottom(botom) {
        this.config.botom = botom;
        return this;
    }

    setRowAround(around) {
        this.config.around = around;
        return this;
    }

    setRowBetween(between) {
        this.config.between = between;
        return this;
    }

    setColWidth(arr) {
        this.setColXS(arr);
        this.setColMD(arr);
        this.setColSM(arr);
        return this.setColLG(arr);
    }

    setColXS(xs) {
        this.config.xs = xs;
        return this;
    }

    setColSM(sm) {
        this.config.sm = sm;
        return this;
    }

    setColMD(md) {
        this.config.md = md;
        return this;
    }

    setColLG(lg) {
        this.config.lg = lg;
        return this;
    }

    setOffset(Offset) {
        this.config.xsOffset = Offset;
        this.config.xsOffset = Offset;
        this.config.lgOffset = Offset;
        this.config.mdOffset = Offset;
        return this;
    }

    setColXSOffset(xsOffset) {
        this.config.xsOffset = xsOffset;
        return this;
    }

    setColSMOffset(smOffset) {
        this.config.xsOffset = xsOffset;
        return this;
    }

    setColMDOffset(mdOffset) {
        this.config.mdOffset = mdOffset;
        return this;
    }

    setColLGOffset(lgOffset) {
        this.config.lgOffset = lgOffset;
        return this;
    }

    setFirst(first) {
        this.config.first = first;
        return this;
    }

    setLast(last) {
        this.config.last = last;
        return this;
    }
}

module.exports = RowColContainer;
