// ui-schema/lib/Table.js
let {WxComponents, WxElementsConstants}     =  require('../constants');
let BaseGroupComponent                      = require('./base/BaseGroupComponent');

class Table extends BaseGroupComponent {

    static  init(param) {
        return new Table(param)
    }

    constructor(param) {
        super(param);
        this.config = {
            'type'      : WxComponents.TABLE,
            showCheckBox: WxElementsConstants.WxTableConstant.SHOW_CHECK_BOX
        };
    }

    setHeaders(headers) {
        this.config.headers = headers;
        return this;
    }

    setTableHeight(bodyHeight) {
        this.config.bodyHeight = bodyHeight;
        return this;
    }

    addHeader(header) {
        if (!this.config.headers)
            this.config.headers = [];
        this.config.headers.push(headers);
        return this;
    }

    setColumns(columns) {
        this.config.columns = columns;
        return this;
    }

    addColumn(header, body, type = 'string') {
        if (!this.config.columns)
            this.config.columns = [];
        this.config.columns.push({header, body, type});
        return this;
    }

    showCheckBoxes() {
        this.config.showCheckBox = true;
        return this;
    }

    hideCheckBoxes() {
        this.config.showCheckBox = false;
        return this;
    }


    setFilterView(filterView) {
        try {
            this.config.filterView = filterView.render();
        } catch (err) {
            this.config.filterView = filterView;
        }
        return this;
    }

    setPaging(currentPage, pageSize, totalPages, serverSide = true) {
        this.config.paging = {
            serverSide,
            currentPage,
            pageSize,
            totalPages
        };
        return this;
    }


}

module.exports = Table;
