// ui-schema/lib/TextField.js
let {WxComponents} =  require('../constants');
let BaseComponent = require('./base/BaseComponent');

class TextField extends BaseComponent {

    static init(params) {
        return new TextField(params);
    }

    constructor(params) {
        super(params);
        this.config = {
            'type': WxComponents.TEXT,

        }
    }

    setValidators(validators) {
        this.config.validators = validators;
        return this;
    }

    setErrorMessages(errorMessages) {
        this.config.errorMessages = errorMessages;
        return this;
    }

    setRightIcon(rightIcon) {
        this.config.rightIcon = rightIcon;
        return this;
    }

    setLeftIcon(leftIcon) {
        this.config.leftIcon = leftIcon;
        return this;
    }

    setType(type) {
        this.config.type = type;
        return this;
    }

    setTypePassword() {
        this.config.input_type = 'password';
        return this;
    }

    setWidthFull() {
        this.config.full = true;
        return this;
    }

    setHint(hint) {
        this.config.hint = hint;
        return this;
    }

    setTitle(title) {
        this.config.title = title;
        return this;
    }

    setDefaultValue(defaultValue) {
        this.config.defaultValue = defaultValue;
        return this;
    }

}


module.exports = TextField;
